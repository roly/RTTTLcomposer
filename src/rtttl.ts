import { Den, NoteEvent, KEYS, KeyDef } from './music';

export interface RTTTLSong {
  name: string;
  defDen: Den;
  bpm: number;
  notes: NoteEvent[];
}

export function parseRTTTL(txt: string, currentDef: Den, currentBpm: number): RTTTLSong {
  const [n, settings, seq] = txt.trim().split(':');
  const parts = settings.split(',');
  let d: Den = currentDef;
  let o = 5;
  let b: number = currentBpm;
  parts.forEach(p => {
    const [k, v] = p.split('=');
    if (k === 'd') d = parseInt(v) as Den;
    if (k === 'o') o = parseInt(v);
    if (k === 'b') b = parseInt(v);
  });
  const evs: NoteEvent[] = [];
  seq.split(',').forEach(tok0 => {
    let tok = tok0.trim();
    if (!tok) return;
    let dot = false;
    if (tok.includes('.')) {
      dot = true;
      tok = tok.replace('.', '');
    }
    let m = tok.match(/^\d+/);
    let den: Den = m ? (parseInt(m[0]) as Den) : d;
    tok = tok.replace(/^\d+/, '');
    if (tok.startsWith('p')) {
      evs.push({ id: crypto.randomUUID(), isRest: true, durationDen: den, dotted: dot });
      return;
    }
    let note = tok[0].toUpperCase();
    tok = tok.slice(1);
    let sharp = false;
    if (tok.startsWith('#')) {
      sharp = true;
      tok = tok.slice(1);
    }
    let oct = o;
    if (tok) oct = parseInt(tok[0]);
    const name = note + (sharp ? '#' : '');
    const keyIndex = KEYS.findIndex((k: KeyDef) => k.name === name && k.octave === oct);
    if (keyIndex >= 0) {
      evs.push({ id: crypto.randomUUID(), isRest: false, keyIndex, note: name, octave: oct, durationDen: den, dotted: dot });
    }
  });
  return { name: n, defDen: d, bpm: b, notes: evs };
}

export function generateRTTTL(name: string, defDen: Den, bpm: number, notes: NoteEvent[]): string {
  const header = `${name}:d=${defDen},o=5,b=${bpm}:`;
  const body = notes
    .map(n => {
      const dur = n.durationDen !== defDen ? n.durationDen.toString() : '';
      const dot = n.dotted ? '.' : '';
      if (n.isRest) return `${dur}p${dot}`;
      return `${dur}${n.note?.toLowerCase()}${KEYS[n.keyIndex!].octave}${dot}`;
    })
    .join(',');
  return header + body;
}

