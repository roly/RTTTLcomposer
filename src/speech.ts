import { NoteEvent, Den, KEYS } from './music';
import { generateRTTTL } from './rtttl';

const STOPWORDS = new Set(['the','a','an','and','or','but','of','to','in','on','for','your']);
const DEG_BY_TIER = [0,2,4,5,7];

export function syllabify(word: string): string[] {
  const clean = word.toLowerCase().replace(/[^a-z]/g,'');
  if(!clean) return [];
  const sylls = clean.match(/[^aeiouy]*[aeiouy]+/g);
  return sylls ?? [clean];
}

export function vowelTier(v: string | undefined): number {
  switch(v){
    case 'u': return 0;
    case 'o': return 1;
    case 'a': return 2;
    case 'e': return 3;
    case 'i':
    case 'y': return 4;
    default: return 2; // mid tier
  }
}

function clamp(n:number,min:number,max:number){
  return Math.max(min,Math.min(max,n));
}

export interface SpeechOptions {
  baseMidi: number;
}

export function speechToEvents(text: string, {baseMidi}: SpeechOptions): NoteEvent[] {
  const lines = text.split(/\n+/).map(l=>l.trim()).filter(Boolean);
  const events: NoteEvent[] = [];
  lines.forEach((line,lineIdx)=>{
    const words = line.split(/\s+/).filter(Boolean);
    words.forEach((word,idxWord)=>{
      const cleanWord = word.toLowerCase().replace(/[^a-z]/g,'');
      const syls = syllabify(word);
      const isStop = STOPWORDS.has(cleanWord);
      syls.forEach((syl,i)=>{
        const v = syl.match(/[aeiouy]/)?.[0];
        const tier = vowelTier(v);
        const degree = DEG_BY_TIER[tier];
        const fall = Math.round((1 - idxWord/Math.max(1, words.length-1))*4) - 2;
        const jitter = i>0? -1: 0;
        const midi = clamp(baseMidi + degree + fall + jitter, 55, 84);
        const key = KEYS.find(k=>k.midi===midi) ?? KEYS[0];
        const stressed = !isStop && i===0;
        let durationDen: Den = stressed ? 4 : 8;
        const isLineFinal = idxWord===words.length-1 && i===syls.length-1;
        if(isLineFinal) durationDen = 4;
        events.push({
          id: crypto.randomUUID(),
          isRest:false,
          keyIndex:key.index,
          note:key.name,
          octave:key.octave,
          durationDen,
        });
      });
      if(/[ktp]$/i.test(cleanWord)){
        events.push({id:crypto.randomUUID(),isRest:true,durationDen:16});
      }
    });
    if(lineIdx<lines.length-1){
      events.push({id:crypto.randomUUID(),isRest:true,durationDen:4});
    }
  });
  return events.slice(0,256);
}

export function speechToRTTTL(name:string, tempo:number, text:string, baseMidi:number){
  const events = speechToEvents(text,{baseMidi});
  const defDen:Den = 8;
  const defOct = 5;
  return generateRTTTL(name, defDen, defOct, tempo, events);
}

