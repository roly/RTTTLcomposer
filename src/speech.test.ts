import { describe, it, expect } from 'vitest';
import { syllabify, vowelTier, speechToEvents } from './speech';
import { generateRTTTL } from './rtttl';

describe('syllabify', () => {
  it('splits words', () => {
    expect(syllabify('pineapple')).toEqual(['pi','nea','pple']);
  });
  it('handles no vowels', () => {
    expect(syllabify('brrr')).toEqual(['brrr']);
  });
});

describe('vowelTier', () => {
  it('maps vowels to tiers', () => {
    expect(vowelTier('u')).toBe(0);
    expect(vowelTier('o')).toBe(1);
    expect(vowelTier('a')).toBe(2);
    expect(vowelTier('e')).toBe(3);
    expect(vowelTier('i')).toBe(4);
    expect(vowelTier('y')).toBe(4);
    expect(vowelTier('x')).toBe(2);
  });
});

describe('speech to RTTTL', () => {
  const baseMidi = 69; // A4
  const phrases = [
    'Hak5 WiFi Pineapple',
    'Trust your technolust',
    'Hack the planet'
  ];
  const tempos = [120, 180];
  phrases.forEach(phrase => {
    tempos.forEach(bpm => {
      it(`${phrase} tempo ${bpm}`, () => {
        const events = speechToEvents(phrase, {baseMidi});
        const rtttl = generateRTTTL('Test',8,5,bpm,events);
        expect(rtttl).toMatchSnapshot();
      });
    });
  });
});
