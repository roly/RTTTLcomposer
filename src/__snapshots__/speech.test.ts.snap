// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`speech to RTTTL > Hack the planet tempo 120 1`] = `"Test:d=8,o=5,b=120:4d#,16p,d,4b4,4b4,16p"`;

exports[`speech to RTTTL > Hack the planet tempo 180 1`] = `"Test:d=8,o=5,b=180:4d#,16p,d,4b4,4b4,16p"`;

exports[`speech to RTTTL > Hak5 WiFi Pineapple tempo 120 1`] = `"Test:d=8,o=5,b=120:4d#,16p,4e,d#,4d,b4,4b4"`;

exports[`speech to RTTTL > Hak5 WiFi Pineapple tempo 180 1`] = `"Test:d=8,o=5,b=180:4d#,16p,4e,d#,4d,b4,4b4"`;

exports[`speech to RTTTL > Trust your technolust tempo 120 1`] = `"Test:d=8,o=5,b=120:4b4,16p,e,4c,g#4,4f#4,16p"`;

exports[`speech to RTTTL > Trust your technolust tempo 180 1`] = `"Test:d=8,o=5,b=180:4b4,16p,e,4c,g#4,4f#4,16p"`;
